File: 0206.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head:
            return ListNode(val='')
        
        head2 = ListNode(head.val)
        head = head.next
        
        while(head):
            head2 = ListNode(val=head.val, next=head2)
            head = head.next
        
        return head2

File: leetcode_75/linked_list/0206.js
//  Definition for singly-linked list.
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  let reverse = (curr, prev) => {
    if (!curr) {
      return null;
    }

    let next = curr.next;
    curr.next = prev;

    if (next) {
      return reverse(next, curr);
    }

    return curr;
  };

  return reverse(head, null);
};

let head = new ListNode(
  1,
  new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, null))))
);
ans = reverseList(head);
console.log(ans);
